### PR TITLE
doc: add RFC for extractors & OpenAPI genration

### PR DESCRIPTION
**Pull request by adriangb**

### CODE REVIEW

File: rfcs/extractors.md [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+## User facing API
+
+The proposed user facing API relies on PEP 593 annotations `file: Annotated[UploadFile, File()]` instead of default values.
+In comparison, FastAPI uses default values (`file: UploadFile = File()`).
+Both of these are just markers: they don't actually do anything except tell the framework how to inject the parameter.
+
+Use of PEP 593 annotations has several advantages:
+
+* Function definitions are not modified at runtime. Functions or other things accepting extracted values can be tested or used as normal functions, including using default values
+* Multiple markers per parameter can be supported, including nested markers
```
Comment (Kludex) [inline rfcs/extractors.md:53]: Remember to highlight those when writing the docs. :+1:

File: rfcs/extractors.md [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+### Data from the request body
+
+```python
+def endpoint(json: AsJson[JsonModel]):
+    ...
+# equivalent to
+def endpoint(json: Annotated[JsonModel, Json()]):
+    ...
+
+def endpoint(file: AsFile[UploadFile]):
```
Comment (Kludex) [inline rfcs/extractors.md:74]: This doesn't match the wording from before `FromFile` and `AsFile`. jfyk

Comment (adriangb) [reply rfcs/extractors.md]: thanks!

**Comment by adriangb** [comment]
This is fully implemented now!

Here's that bakers complex case of nested form data (a form where the value is itself a form):

https://github.com/adriangb/rapi/blob/4a91ec95cd717beb7199a4f93376f5180402a9e3/tests/test_extractors/body/test_form_data_extractor.py#L223-L238

This is about as complicated as it gets.
I think it's pretty cool that in ~5 LOC you can abstract away hundreds of complex LOC of parsing, validation, concurrency, etc.

**Pull request closed without merging**